### PR TITLE
Pub 945 collate backend blobs

### PR DESCRIPTION
**Pull request by furnivall**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/PUB-945

### Change description ###
Added two new GET endpoints:

####  '/search' #### 
This retrieves from Postgres all data fulfilling the following criteria:
-  A court id exists within the "search" criteria (populated via json path)
-  The display-from value is before the current date time and display-to date after.
-  If the "verified" boolean is true, then the search will return all sensitivity types. 
- If false, then it will only return "public" sensitivity data.

#### '/blob' ####
-  this retrieves the data within a blob as a String from the azure blob storage based on a given URI (which is made up of a combination of the provenance and sourceArtefactID fields.

### Integration test issues: ###
Due to the use of Postgres in real-world usage and H2 in-memory db for integration tests, this PR cannot be currently integration tested (Postgres-specific syntax is used which cannot be emulated by h2). This will be amended in a future ticket, where we will aim to bring in local Postgres for testing.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X ] No
```

### CODE REVIEW

**Review by kurt-uk — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 1]
```diff
@@ -184,6 +184,8 @@ dependencies {
 
   implementation 'org.springframework.boot:spring-boot-starter-data-jpa:2.5.5'
   implementation('org.postgresql:postgresql')
+    implementation 'junit:junit:4.13.1'
```
[the remark this answers is not in the document]
Comment (kurt-uk) [reply build.gradle:187]: (all the junit 4 ones)

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/errorhandling/GlobalExceptionHandler.java [hunk 2]
```diff
@@ -74,4 +77,19 @@
         return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(exceptionResponse);
     }
 
+    @ExceptionHandler(BlobStorageException.class)
+    public ResponseEntity<ExceptionResponse> handle(BlobStorageException ex) {
+        ExceptionResponse exceptionResponse = new ExceptionResponse();
+        if (Objects.equals(ex.getErrorCode().toString(), "BlobNotFound")) {
+            exceptionResponse.setMessage("404: Unable to find a blob matching the given inputs");
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/errorhandling/GlobalExceptionHandler.java:84]: I would avoid relying on Strings to determine which route to go down (as this would not be picked up at compilation time if Azure were to change the string in the future). 

This should probably be two different Exceptions, captured and then thrown at the source of where they are occurring.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java [hunk 3]
```diff
@@ -32,25 +34,60 @@ public PublicationService(ArtefactRepository artefactRepository,
...
+    }
+
+    /**
+     * takes in provenance and artefact id and returns the data within the matching blob in string format.
+     *
+     * @param provenance represents the provenance component of the blob
+     * @param artefactId represents the artefact id component of the blob
+     * @return the data within the blob in string format
+     */
+    public String getFromBlobStorage(String provenance, String artefactId) {
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:88]: Similar to the API comment around sourceArtefactID in the controller. I would rename this to 'getByArtefactId', as it then unlinks the calling classes (e.g controller) from the underlying repository (blob)

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java [hunk 4]
```diff
@@ -32,25 +34,60 @@ public PublicationService(ArtefactRepository artefactRepository,
 
     /**
      * Method that handles the creation or updating of a new publication.
+     *
      * @param artefact The artifact that needs to be created.
-     * @param payload The paylaod for the artefact that needs to be created.
+     * @param payload  The paylaod for the artefact that needs to be created.
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:39]: Not part of this PR :) But while you're in this file, could you update the typo in payload

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (furnivall) [reply build.gradle:187]: So this was because I wanted to @ignore an integration test until Postgres was fixed. turns out @ignore is the junit 4 equivalent of @disabled - should now be fixed :)

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java [hunk 5]
```diff
@@ -87,6 +91,22 @@ public PublicationController(PublicationService publicationService) {
         return ResponseEntity.status(HttpStatus.CREATED).body(createdItem);
     }
 
+    @ApiOperation("Get a series of publications matching a given input (e.g. courtid)")
```
[the remark this answers is not in the document]
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java:94]: Should be sorted now :)

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java [hunk 6]
```diff
@@ -87,6 +91,22 @@ public PublicationController(PublicationService publicationService) {
...
+    public ResponseEntity<List<Artefact>> getAllRelevantArtefacts(@RequestHeader String searchValue,
+                                                                   @RequestHeader Boolean verification) {
+        return ResponseEntity.status(HttpStatus.OK)
+            .body(publicationService.findAllWithSearch(searchValue, verification));
+    }
+
+    @ApiOperation("Get the info from within a blob given source artefact id and provenance")
+    @GetMapping("/blob")
+    public ResponseEntity<String> getBlobData(@RequestHeader String sourceArtefactId, String provenance) {
+        return ResponseEntity.status(HttpStatus.OK)
```
[the remark this answers is not in the document]
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java:105]: Fixed as above.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java [hunk 7]
```diff
@@ -87,6 +91,22 @@ public PublicationController(PublicationService publicationService) {
         return ResponseEntity.status(HttpStatus.CREATED).body(createdItem);
     }
 
+    @ApiOperation("Get a series of publications matching a given input (e.g. courtid)")
+    @GetMapping("/search")
+    public ResponseEntity<List<Artefact>> getAllRelevantArtefacts(@RequestHeader String searchValue,
```
[the remark this answers is not in the document]
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java:96]: Sorry! Should be fixed now.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java [hunk 8]
```diff
@@ -87,6 +91,22 @@ public PublicationController(PublicationService publicationService) {
...
+                                                                   @RequestHeader Boolean verification) {
+        return ResponseEntity.status(HttpStatus.OK)
```
[the remark this answers is not in the document]
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java:98]: Fixed now.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java [hunk 9]
```diff
@@ -32,25 +34,60 @@ public PublicationService(ArtefactRepository artefactRepository,
...
+    /**
+     * Get all relevant artefacts relating to a given court ID.
+     *
+     * @param searchValue - represents the court ID in question being searched for
+     * @param verified    - represents the verification status of the user. Currently only verified/non-verified, but
+     *                    will include other verified user types in the future
+     * @return a list of all artefacts that fulfil the timing criteria, match the given court id and sensitivity
+     *                    associated with given verification status
+     */
+    public List<Artefact> findAllWithSearch(String searchValue, Boolean verified) {
```
[the remark this answers is not in the document]
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:71]: fixed :)

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java [hunk 10]
```diff
@@ -87,6 +91,22 @@ public PublicationController(PublicationService publicationService) {
...
+            .body(publicationService.findAllWithSearch(searchValue, verification));
+    }
+
+    @ApiOperation("Get the info from within a blob given source artefact id and provenance")
```
[the remark this answers is not in the document]
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java:102]: fixed

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java (hunk 4 above)
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:39]: Sorted :)

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java (hunk 3 above)
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:88]: Done.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java [hunk 11]
```diff
@@ -87,6 +91,22 @@ public PublicationController(PublicationService publicationService) {
...
+    @GetMapping("/blob")
```
[the remark this answers is not in the document]
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java:103]: Hi @ChrisS1512 - should artefactId be a path param or a header? I guess it's kind of like a search so to be consistent with Kurt's comments above would make sense to put it as that?

File: src/test/java/uk/gov/hmcts/reform/pip/data/management/database/AzureBlobServiceTest.java [hunk 12]
```diff
@@ -40,4 +44,15 @@ void testCreationOfNewBlob() {
         assertEquals(CONTAINER_URL + "/" + blobName, blobUrl, "Payload URL does not"
             + "contain the correct value");
     }
+
+    @Test
+    void testGetBlobData() {
+        String blobName = SOURCE_ARTEFACT_ID + '-' + PROVENANCE;
+        when(blobContainerClient.getBlobClient(blobName)).thenReturn(blobClient);
+        doNothing().when(blobClient).downloadStream(any());
+        assertThat("Error Message", azureBlobService.getBlobData(SOURCE_ARTEFACT_ID, PROVENANCE),
```
[the remark this answers is not in the document]
Comment (furnivall) [reply src/test/java/uk/gov/hmcts/reform/pip/data/management/database/AzureBlobServiceTest.java:53]: sorted.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/database/ArtefactRepository.java [hunk 13]
```diff
@@ -1,13 +1,29 @@
...
+import java.time.LocalDateTime;
+import java.util.List;
 import java.util.Optional;
 
 @Repository
 public interface ArtefactRepository extends JpaRepository<Artefact, Long> {
     Optional<Artefact> findBySourceArtefactIdAndProvenance(String sourceArtefactId, String provenance);
 
+
+    @Query(value = "select * from Artefact where search->'court-id'->>0 = :searchVal and display_from < :curr_date "
```
[the remark this answers is not in the document]
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/database/ArtefactRepository.java:20]: image
We've done some deep digging on a local instance of Postgres where we could look at logs and it seems okay, as it's using parametrised queries: 
image

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/database/ArtefactRepository.java [hunk 14]
```diff
@@ -1,13 +1,29 @@
...
+        + "and display_to> :curr_date",
+        nativeQuery = true)
+    List<Artefact> findArtefactsBySearchVerified(@Param("searchVal") String searchVal,
+                                         @Param("curr_date") LocalDateTime currentDate);
+
+    @Query(value = "select * from Artefact where search->'court-id'->>0 = :searchVal and sensitivity = 'PUBLIC' and "
```
[the remark this answers is not in the document]
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/database/ArtefactRepository.java:26]: see above response.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/errorhandling/GlobalExceptionHandler.java [hunk 15]
```diff
@@ -74,4 +77,19 @@
         return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(exceptionResponse);
     }
 
+    @ExceptionHandler(BlobStorageException.class)
+    public ResponseEntity<ExceptionResponse> handle(BlobStorageException ex) {
```
[the remark this answers is not in the document]
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/errorhandling/GlobalExceptionHandler.java:81]: So for this one, the BlobStorageException is built into Azure, not something I wrote. 

I deliberately separated out the 404 because I could reproduce a 404 on demand. I wasn't sure how to create any other type of BlobStorageException on demand, but I didn't want them erroneously characterised as a 404. 

In response to Chris's comment below, I've stopped using the string value comparison and have used a BlobErrorCode comparison instead, which should be more stable.

Does this make sense?

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/errorhandling/GlobalExceptionHandler.java (hunk 2 above)
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/errorhandling/GlobalExceptionHandler.java:84]: Solved :)

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ChrisS1512) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java:100]: Might need to add the error states to this (might just be the 404)

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java [hunk 16]
```diff
@@ -32,25 +35,67 @@ public PublicationService(ArtefactRepository artefactRepository,
...
+        if (verified) {
+            return artefactRepository.findArtefactsBySearchVerified(searchValue, currDate);
+        } else {
+            return artefactRepository.findArtefactsBySearchUnverified(searchValue, currDate);
+        }
+
+    }
+
+    /**
+     * takes in provenance and artefact id and returns the data within the matching blob in string format.
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:83]: This java doc description needs updating now that just the artefact ID is passed in

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java [hunk 17]
```diff
@@ -32,25 +35,67 @@ public PublicationService(ArtefactRepository artefactRepository,
...
+    public String getByArtefactId(UUID artefactId) {
+        Optional<Artefact> optionalArtefact = artefactRepository.findByArtefactId(artefactId);
+        if (optionalArtefact.isPresent()) {
+            Artefact artefact;
+            artefact = optionalArtefact.get();
+            String sourceArtefactId = artefact.getSourceArtefactId();
+            String provenance = artefact.getProvenance();
+            return azureBlobService.getBlobData(sourceArtefactId, provenance);
+        } else {
+            return azureBlobService.getBlobData("", "");
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:97]: Is this needed / can we just throw an exception here instead?

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (ChrisS1512) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java]: Search should be a path param I think, and then verified be the header :)

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java (hunk 17 above)
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:97]: To be honest, I just did this so as not to break the existing tests. Absolutely can just throw an exception, yeah :)

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java]: Done

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java (hunk 16 above)
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationService.java:83]: Sorted.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (furnivall) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java:102]: Should now be done.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java [hunk 18]
```diff
@@ -87,6 +93,32 @@ public PublicationController(PublicationService publicationService) {
...
+    @ApiResponses({
+        @ApiResponse(code = 200,
+            message = "Blob data from the given request in text format.",
+            response = String.class),
+        @ApiResponse(code = 404,
+            message = "No artefact found matching given parameters and date requirements"),
+    })
+    @ApiOperation("Get the info from within a blob given source artefact id and provenance")
+    @GetMapping("/{artefactId}")
+    public ResponseEntity<String> getBlobData(@PathVariable UUID artefactId) {
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/PublicationController.java:118]: Just been trying this out. Looks good! I think we'll want the verified flag on the getBlobData endpoint too in the header, just to make sure that the user requesting the data from the blob is verified

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/database/ArtefactRepository.java [hunk 19]
```diff
@@ -1,13 +1,31 @@
...
 import java.util.Optional;
+import java.util.UUID;
 
 @Repository
 public interface ArtefactRepository extends JpaRepository<Artefact, Long> {
     Optional<Artefact> findBySourceArtefactIdAndProvenance(String sourceArtefactId, String provenance);
 
+    Optional<Artefact> findByArtefactId(UUID artefactId);
+
+    @Query(value = "select * from Artefact where search->'court-id'->>0 = :searchVal and display_from < "
```
Comment (ChrisS1512) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/database/ArtefactRepository.java:20]: Just been testing it all out. Display from and display to are optional fields, therefore if they are not set no results are returned.

I'm not sure what the expected behaviour for no display from / to is, so we'll need to double check this.

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]

**Review by ChrisS1512 — DISMISSED** [review]
LGTM

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/database/ArtefactRepository.java (hunk 19 above)
Comment (ChrisS1512) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/database/ArtefactRepository.java:20]: This will now be captured in a seperate ticket

**Review by marcogagliardi — APPROVED** [review]
👌🏽

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/database/ArtefactRepository.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (kurt-uk) [reply src/main/java/uk/gov/hmcts/reform/pip/data/management/database/ArtefactRepository.java]: ok thats good then! can you pop a couple of unit tests in that prove that too?

File: src/test/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationServiceTest.java [hunk 20]
```diff
@@ -121,4 +126,67 @@ void testUpdatingOfExistingArtefact() {
...
+
+        when(artefactRepository.findByArtefactId(any())).thenReturn(Optional.of(artefact));
+        assertThrows(UnauthorisedRequestException.class, () -> {
+            publicationService.getByArtefactId(UUID.randomUUID(), false);
+        }, "Should throw an unauthorised request exception.");
+    }
 }
+
+
+
```
Comment (kurt-uk) [inline src/test/java/uk/gov/hmcts/reform/pip/data/management/service/PublicationServiceTest.java:192]: image

**Review by kurt-uk — CHANGES_REQUESTED** [review]

**Review by kurt-uk — APPROVED** [review]
